libass/ass_render.c

```diff
@@ -1010,21 +1010,23 @@ static void init_font_scale(ASS_Renderer *render_priv)
     render_priv->screen_scale_y = font_scr_h / render_priv->track->PlayResY;
 
     ASS_Vector layout_res = ass_layout_res(render_priv);
-    render_priv->blur_scale = font_scr_h / layout_res.y;
+    render_priv->blur_scale_x = font_scr_w / layout_res.x;
+    render_priv->blur_scale_y = font_scr_h / layout_res.y;
```

Comment:
### `LayoutRes` vs `ass_set_pixel_aspect` for blur and 3D transforms:

Normally, `(font_scr_w / layout_res.x) == par_scale_x * (font_scr_y / layout_res.y)`. Indeed, when exactly one of `par_scale_x` and `layout_res` is known, the other is defaulted to a value that ensures this equation is satisfied.

But our API allows the consumer to explicitly set _both_ `storage_size` and `pixel_aspect` to such values that this equation doesn’t hold. In practice, this happens (only) in mpv with `--sub-ass-vsfilter-blur-compat=yes --sub-ass-vsfilter-aspect-compat=no`.

In such a situation, it’s anyone’s guess as to what “should” happen in the user’s opinion. The only scenarios I’ve been able to come up with when a user might want to do this involve mixed-and-match video and subtitles:

* Maybe they’re playing an anamorphic video file with custom style overrides or converted subtitles and seeing that without `aspect-compat=no` their overridden style is getting stretched ugly.

    In this case, it’s obvious we shouldn’t be stretching any text (so we don’t want to disregard the PAR completely as we [tried](https://github.com/libass/libass/pull/645#pullrequestreview-1119327502)), but if there’s any blur or transforms left from the original ASS, it may make more sense still to apply them at the anamorphically-compressed LayoutRes, as originally intended, despite the `aspect-compat=no`.

     This PR currently does this for blur, but transforms are applied (in master, and this PR doesn’t currently change this) to an anamorphically-stretched video-height rectangle. Maybe we ought to change the transform code, too.

    Then again, user style overrides can include `Blur` and `Angle`, too, and in that case it’d make more sense to apply them at the display aspect ratio. Maybe we ought to adjust user overrides for `par_scale_x` in `handle_selective_style_overrides` like we adjust them for `PlayResY`?

* Maybe they’re playing an anamorphic video file with an external subtitle file that was authored for 
an unrelated non-anamorphic video file, as in https://github.com/mpv-player/mpv/issues/10680#issuecomment-1257127239.

    In this case, well, what the user really wants us to do is to pretend `layout_res` itself has the anamorphically-stretched aspect ratio. Indeed, this is what we assume with `blur-compat=no aspect-compat=no` (and before b6db99a3930b609ca1ba5a0b9930b5c6dc5701dc assumed with `blur-compat=no` regardless of `aspect-compat`). But at the end of the day, `blur-compat=no` is the _only_ way to handle this situation: if the external subs don’t declare `LayoutRes`, they may have been authored for _any_ video, and the current video’s resolution is completely irrelevant.

Reply:
> [...] or converted subtitles and seeing that without aspect-compat=no their overridden style is getting stretched ugly.

For converted subs storage resolution is never set and pixel aspect is at most set to the *display’s* pixel aspect.

> Then again, user style overrides can include `Blur` and `Angle`, too, and in that case it’d make more sense to apply them at the display aspect ratio. Maybe we ought to adjust user overrides for par_scale_x in handle_selective_style_overrides like we adjust them for PlayResY?

I can see how this could work for `Angle`, but I’m not sure how you want to adjust the single `Blur` and in fact also `Shadow` and `Outline` values of a Style to avoid stretching in one direction. Do you want to involve the renderer state with its split `(x|y)(shad|bord)` values, override the scaling factors for the current event (also affecting any original ASS snippets) or something else? *(+ with SBAS=yes Shadow and Outline would need to be treated differently depending on whether PlayRes* matches storage, display res or none of the two)*

Reply:
Right; I didn’t actually think about it. But it looks simple enough: currently, we take user overrides in `handle_selective_style_overrides` and put them in a `ASS_Style`, which lacks the split x/y values. But the only place that calls `handle_selective_style_overrides` immediately converts that style to the `RenderContext`, which does have them. So perhaps we could make `handle_selective_style_overrides` just modify the context directly. It looks like the only part of the `state.style` that isn’t just copied to another field is the `Margin{L,R,V}`: perhaps we should just give them their own fields in the `RenderContext`.

Ironically, I’m _less_ clear on `Angle` now that I think about this. Normally, `Angle` means `\frz`. For one, if the user overrides `\frz`, do they also want to override `\frx0\fry0`? Even if so, `\frz` followed by horizontal stretching of the anamorphic video is no longer representable as a single rotation: we’d need to decompose it into multiple values. Or just create an extra flag and set it to disable `par_scale_x` for this one glyph run… ew.

Besides, we can never be quite sure what the right way to “disable” anamorphic compensation is. mpv apparently supports rendering to non-square-pixel displays and sets `par` appropriately. But we can’t distinguish the display PAR from the video PAR, so we can’t compensate overrides for just one of the two PARs. Ugh.

On another note, though, even if not for PAR, we probably totally should compensate horizontal-measure overrides for `PlayResX` now that we scale them. In particular, yes, `\xbord` and `\xshad`, so we need to apply overrides directly to `RenderContext` anyway.

> override the scaling factors for the current event

wouldn’t affect borders and shadows.

> with SBAS=yes Shadow and Outline would need to be treated differently depending on whether PlayRes matches storage, display res or none of the two

Not sure I understand.

Reply:
> > override the scaling factors for the current event
> wouldn’t affect borders and shadows.

I was referring to `border_scale_(x|y)`, `blur_scale_(x|y)`, etc

> > with SBAS=yes Shadow and Outline [...]
> Not sure I understand.

Nvm, I got something wrong.